src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newStates = [Object.assign({}, state)];
+
+  for (const action of actions) {
+    const type = action.type;
+    let newState = Object.assign({}, newStates[newStates.length - 1]);
+
+    switch (type) {
+      case 'addProperties': {
+        newState = Object.assign(newState, action.extraData);
+        break;
+      }
+
+      case 'removeProperties': {
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+      }
+
+      case 'clear': {
```

Comment:
Not critical, but would be great to use variables instead of the hardcoded values